ddl/check.lua

```diff
@@ -628,75 +628,147 @@ local function check_index(i, index, space)
-                " no bucket_id defined in 'format' section",
-                space.name
-            )
-        end
+    -- all objects with type `cdata` are allowed
+    -- because there is no easy way to get
+    -- metatable.__call of object with type `cdata`
+    if type(object) == 'cdata' then
+        return true
+    end
 
-        if space.fields.bucket_id.type ~= 'unsigned' then
-            return nil, string.format(
-                "spaces[%q].format[%q].type: bad value (unsigned expected, got %s)",
-                space.name, 'bucket_id', space.fields.bucket_id.type
-            )
-        end
+    local object_metatable = getmetatable(object)
+    if (type(object) == 'table' or type(object) == 'userdata')
+    and object_metatable and object_metatable.__call then
+        -- `__call` metamethod can be only the `function`
+        -- and cannot be a `table` | `userdata` | `cdata`
+        -- with `__call` methamethod on its own
+        return type(object_metatable.__call) == 'function'
+    end
+
+    return false
+end
+
+local function check_custom_sharding_func(sharding_func_name)
+    local sharding_func = rawget(_G, sharding_func_name)
+    if sharding_func == nil then
+        return false
+    end
+
+    return is_callable(sharding_func)
+end
+
+local function check_sharding_func(space)
+    if not space.sharding_func then
+        return true
     end
 
-    do -- check that indexes['bucket_id'] valid
-        if not space.indexes.bucket_id then
+    do -- check that sharding_func format is valid
+        if type(space.sharding_func) ~= 'string' and type(space.sharding_func) ~= 'table' then
             return nil, string.format(
-                "spaces[%q].indexes: sharding_key exists in the space, but there's" ..
-                " no bucket_id defined in 'indexes' section",
-                space.name
+                "spaces[%q].sharding_func: bad value" ..
+                " (string or table expected, got %s)",
+                space.name, type(space.sharding_func)
             )
         end
 
-        if space.indexes.bucket_id.unique then
-            return nil, string.format(
-                "spaces[%q].indexes[%q].unique: bucket_id index can't be unique",
-                space.name, 'bucket_id'
-            )
+        if type(space.sharding_func) == 'string' then
+            -- sharding func name in dot notation must be
+            -- sequence of letters, digits, or underscore symbols,
+            -- and may not begin with a digit or dot, the last symbol
+            -- may not be a dot
+            if not space.sharding_func:match('^[%a_]%w*[%.[%w_]+%.]*[%w_]+$')
+            and not check_custom_sharding_func(space.sharding_func) then
+                return nil, string.format(
+                    "spaces[%q].sharding_func: unknown sharding_func %q",
+                    space.name, space.sharding_func
+                )
+            end
```

Comment:
(I wrote the text below during thinking, so a lot of thoughts around are there. You can skip it, the summary is at the bottom.)

First of all, it seems, the regex is broken by square brackets within square brackets:

```lua
tarantool> ('xxx'):match('^[%a_]%w*[%.[%w_]+%.]*[%w_]+$')
---
- null
...
```

Or we check one component names separately and call them 'custom'? Consider more examples:

```lua
tarantool> ('foo..bar'):match('^[%a_]%w*[%.[%w_]+%.]*[%w_]+$')
---
- foo..bar
...
tarantool> ('foo.ba[r.baz'):match('^[%a_]%w*[%.[%w_]+%.]*[%w_]+$')
---
- foo.ba[r.baz
...
```

(It seems that Lua regexps are not powerful enough for validation of dot notation, see below.)

Next, LuaJIT accepts some ASCII symbols and any correct UTF-8 multibyte sequences (and some incorrect ones) as indentifier symbols. The token should not start from a digit and should not be a reserved word (like 'end'). See `lj_char_isident()`, `lj_lex()` and `TKDEF` macro at top of the `lj_lex.h` file.

Tarantool requires correct printable UTF-8 characters in `box.schema.func.create()`, but not necessarily an identifier.

See examples:

```lua
$ echo -en 'local function \xc0\x61\x80\x61(...) print(...) end \xc0\x61\x80\x61(1,2,3)' | tarantool
1	2	3
```

```lua
tarantool> ok, err = pcall(box.schema.func.create, '\xc0\x61\x80\x61')
tarantool> print(tostring(err))
Invalid identifier 'Àaa' (expected printable symbols only or it is too long)
```

But if fact we're not restricted by correct indentifiers:

```lua
tarantool> _G['foo bar'] = function(...) print(...) end
tarantool> _G['1'] = function(...) print(...) end
tarantool> _G['end'] = function(...) print(...) end

tarantool> _G['foo bar'](1, 2, 3)
1	2	3
tarantool> _G['1'](1, 2, 3)
1	2	3
tarantool> _G['end'](1, 2, 3)
1	2	3

tarantool> box.cfg{listen = 3301}
tarantool> box.schema.user.grant('guest','read,write,execute,create,drop,alter','universe')
tarantool> net_box = require('net.box')
tarantool> c = net_box.connect(3301)

tarantool> c:call('foo bar', {1, 2, 3})
1	2	3
tarantool> c:call('1', {1, 2, 3})
1	2	3
tarantool> c:call('end', {1, 2, 3})
1	2	3
```

(`box.schema.func.create()` accepts such non-indentifier names too.)

Iproto calls and `box.func.<name>.call()` also accept a colon:

```
tarantool> _G.foo = {}
tarantool> _G.foo.bar = function(...) print(...) end
tarantool> box.schema.func.create('foo:bar')

tarantool> box.func['foo:bar']:call({1,2,3})
table: 0x40b19da0	1	2	3
tarantool> c:call('foo:bar', {1, 2, 3})
table: 0x40b19da0	1	2	3
```

(Locally we can even use non-string keys in `_G`.)

To be honest, I don't know what is better to do here. I don't like artificial restrictions. However I doubt that non identifier names are really used anywhere (as well as non ASCII ones).

Our purpose it to protect a user from an obvious mistake. What kind of errors we expect? Non-string value is likely a mistake. Non ASCII names too. Non identifier names as well.

So it looks practical to follow LuaJIT's rules (or even restrict them to ASCII):

1. Split an identifier by dots.
2. Verify those chunks.

At least, the chunks should NOT:

1. Be empty.
2. Be equal to a reserved keyword.
3. Contain control, space, punctuation symbols: we can grab LuaJIT's table
   from `lj_char.c` and use the `bit` module to reimplement `lj_char_isident`
   macro. Or use a Lua regexp for a chunk.
4. Start from a digit.

We can express the rules using a regexp (but not Lua regexp, see below). `lj_char_isident()` accepts the following ASCII characters: `[a-zA-Z0-9_]` (or `[%w_]`). So (if we restrict users to ASCII) the following regex would be correct: `^[%a_](%.[%w_]+)*$` if Lua would support repeating captures. It seems impossible to form a Lua regex that would accept 'foo.bar' like strings and will not accept 'foo..bar' for arbitrary number of chunks.

So we should actually split the identifier by dots and check the first and next chunks using regexps or using a symbol classes table (like `lj_char_bits`).

----

For the sake of completeness, there is a relevant issue in tarantool: https://github.com/tarantool/tarantool/issues/3405.

----

The code to check how `lj_char_isident()` and other macros work:

```
tarantool> lj_char_bits = {0, <...>} -- copy from lj_char.c
tarantool> function lj_char_isident(n)
               return bit.band(lj_char_bits[n + 2], 0x80) == 0x80
           end
tarantool> for i = 0, 255 do print(i, string.char(i), lj_char_isident(i)) end
```

Reply:
Regexp way would be simpler (if we're about a chunk validation), but I don't mind the LuaJIT way (it gains some UTF-8 support). If there will be other things to discuss around the dot syntax validation code, I'll add my questions as separate comments. I'm going to mark this thread as resolved.

Reply:
Thanks for the update!